### PR TITLE
ListDataView: addItemAfter doesn't work with filter

### LINKED ISSUE #8762: ListDataView: addItemAfter doesn't work with filter
Without filter, `addItemAfter` works nicely. But after applying a filter, the item could be added before the target.

To reproduce:
1. Select the last row `AAA` in the grid
2. Click the `Add a new item` button, the new item `BBB` is added correctly after `AAA`. 
3. So far so good, now refresh the page to start again
4. Click on the `Add filter` button
5. Select the last row `AAA` in the grid
6. Click the `Add a new item` button, the new item `BBB` is added *before* `AAA`

Code to reproduce:
```
public class DashboardView extends Div {

    private Grid<Person> grid = new Grid<Person>(Person.class);

    public static class Person {
        private String firstName, lastName;

        public Person(String firstName, String lastName) {
            this.firstName = firstName;
            this.lastName = lastName;
        }

        public String getFirstName() {
            return firstName;
        }

        public void setFirstName(String firstName) {
            this.firstName = firstName;
        }

        public String getLastName() {
            return lastName;
        }

        public void setLastName(String lastName) {
            this.lastName = lastName;
        }

    }

    public DashboardView() {
        setId("dashboard-view");
        add(new Button("Add a new item", e->addNewItemAfterSelection()));
        add(new Button("Add filter", e->addFilter()));
        add(new Button("Remove filter", e->removeFilter()));
        
        List<Person> persons = new ArrayList<Person>();
        persons.addAll(Arrays.asList(new Person("John", "Doe"), new Person("Jane", "Jo"), new Person("AAA", "AAA")));
        grid.setItems(persons);
        add(grid);
    }

    private void addFilter(){
        grid.getListDataView().addFilter(person -> !person.getLastName().equals("Doe"));
    }

    private void removeFilter(){
        grid.getListDataView().removeFilters();
    }

    private void addNewItemAfterSelection(){
        Person currentPerson = grid.asSingleSelect().getValue();
        if(currentPerson!=null){
            grid.getListDataView().addItemAfter(new Person("BBB", "BBB"), currentPerson);
        }
    }
}
```

### PR DESCRIPTION
**Pull request by mshabarov**
Fixes #8762

Thanks @mstahv for this fix! I've just added a regression tests and created a new branch.

### CODE REVIEW

File: flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java [hunk 1]
```diff
@@ -267,11 +268,11 @@ protected void validateItemIndex(int itemIndex) {
         return this;
     }
 
-    private int getItemIndex(T item) {
+    private int getItemIndex(T item, Stream<T> stream) {
         Objects.requireNonNull(item, NULL_ITEM_ERROR_MESSAGE);
         AtomicInteger index = new AtomicInteger(-1);
         //@formatter:off
-        if (!getItems().peek(t -> index.incrementAndGet())
+        if (!stream.peek(t -> index.incrementAndGet())
```
Comment (denis-anisimov) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java:275]: Please avoid one letter variable

Comment (denis-anisimov) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java:275]: Also : it's an existing code but I'm a bit uncomfortable with using `peek` which should not have side effects but it has here.
This is minor.

Comment (mshabarov) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java]: > Please avoid one letter variable

Done.

> peek which should not have side effects but it has here

Well, this was done as a simple solution to count the items while the other terminate method is expected (`findFirst()`). The side effect here (counter increment) is intend. I don't know how to do that in more elegant way.

Comment (denis-anisimov) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java]: There is always a way to use `forEach`  and implement the code in other way.
But it's minor as I said.

**Pull request merged**